tests/unit/fem/test_quadf_coef.cpp

```diff
@@ -0,0 +1,197 @@
+      {
+         std::cout << "  Testing GridFunc L2 projection" << std::endl;
+         L2_FECollection    fec_l2(order_h1, dim);
+         FiniteElementSpace fespace_l2(&mesh, &fec_l2, dim);
+         GridFunction g0(&fespace_l2);
+         GridFunction gtrue(&fespace_l2);
+
+         {
+            int nnodes = gtrue.Size() / dim;
+            int vdim = dim;
+
+            for (int i = 0; i < vdim; i++)
+            {
+               for (int j = 0; j < nnodes; j++)
+               {
+                  gtrue((i * nnodes) + j) = i;
+               }
+            }
+         }
+
+         g0 = 0.0;
+         g0.ProjectDiscCoefficient(qfvc, GridFunction::ARITHMETIC);
+         gtrue -= g0;
+         REQUIRE(gtrue.Norml2() < tol);
+      }
+
+      SECTION("Gridfunction H1 tests")
+      {
+         std::cout << "  Testing GridFunc H1 projection" << std::endl;
+         H1_FECollection    fec_h1(order_h1, dim);
+         FiniteElementSpace fespace_h1(&mesh, &fec_h1, dim);
+         GridFunction g0(&fespace_h1);
+         GridFunction gtrue(&fespace_h1);
+
+         {
+            int nnodes = gtrue.Size() / dim;
+            int vdim = dim;
+
+            for (int i = 0; i < vdim; i++)
+            {
+               for (int j = 0; j < nnodes; j++)
+               {
+                  gtrue((i * nnodes) + j) = i;
+               }
+            }
+         }
+
+         g0 = 0.0;
+         g0.ProjectCoefficient(qfvc);
+         gtrue -= g0;
+         REQUIRE(gtrue.Norml2() < tol);
+      }
+   }
+
+   SECTION("Operators on QuadFuncCoeff")
+   {
+      SECTION("Gridfunction L2 tests")
+      {
+         std::cout << "Testing QuadFuncCoeff:";
+         std::cout << "  Testing GridFunc L2 projection" << std::endl;
+         L2_FECollection    fec_l2(order_h1, dim);
+         FiniteElementSpace fespace_l2(&mesh, &fec_l2, 1);
+         GridFunction g0(&fespace_l2);
+         GridFunction gtrue(&fespace_l2);
+
+         {
+            int nnodes = gtrue.Size();
+            int vdim = 1;
+
+            for (int i = 0; i < vdim; i++)
+            {
+               for (int j = 0; j < nnodes; j++)
+               {
+                  gtrue((i * nnodes) + j) = 1.0;
+               }
+            }
+         }
+
+         g0 = 0.0;
+         g0.ProjectDiscCoefficient(qfc, GridFunction::ARITHMETIC);
+         gtrue -= g0;
+         REQUIRE(gtrue.Norml2() < tol);
+      }
+
+      SECTION("Gridfunction H1 tests")
+      {
+         std::cout << "  Testing GridFunc H1 projection" << std::endl;
+         H1_FECollection    fec_h1(order_h1, dim);
+         FiniteElementSpace fespace_h1(&mesh, &fec_h1, 1);
+         GridFunction g0(&fespace_h1);
+         GridFunction gtrue(&fespace_h1);
+
+         {
+            int nnodes = gtrue.Size();
+            int vdim = 1;
+
+            for (int i = 0; i < vdim; i++)
+            {
+               for (int j = 0; j < nnodes; j++)
+               {
+                  gtrue((i * nnodes) + j) = 1.0;
```

Comment:
Could we do a non-constant example?

Reply:
I don't see why not. I don't know maybe something like where we have the quadrature point set to their physical space z values. If my understanding of the ```GridFunction::Project``` is right and it's doing something akin to a L2 projection method I would expect the projected gridfunction to have roughly the same values as the mesh's z values.

Reply:
So, I've been recently been adding the tests in for this, and it doesn't appear to me that ```GridFunction::Project``` or ```GridFunction::DiscProject``` as far as I can tell aren't doing an L2 type projection. It appears that for ```GridFunction::DiscProject``` the nodes simply become what values are at the quadrature points. I'll add a commits with these test even though the H1 project fails.

Reply:
Yes, I think we need to handle these quadrature coefficients as a special cases in the `Project` methods. They can call separate new method, say `(Par)GridFunction::ProjectQuadratureFunction`, that performs an L2-projection given a `QuadratureFunction`.

Reply:
@v-dobrev I've done some work with L2-projection methods in python before. 

I mainly dealt with two different formulations one where you solved the following system:
[N_sf]{x_nodal_pts} = {quad_pts}
where [N_sf] can be inferred as a global shape function matrix that maps the nodal points to the quadrature points. This formulation required a least squares solver to solve for the nodal points. One nice benefit of this one though is I feel it could be used to project the data down to a lower order finite element space, since we would essentially be just solving an over-determined system at that point.
The other formulation was the more classical one seen in https://doi.org/10.1016/0168-874X(94)00054-J and is Eq 13. We should be able to come up with a partial assembly formulation for this one, so it can run on the GPU. I'd like to also do this for a ProjectDiscCoeff method which I'm going guessing Yohann's work with the DG PA formulation would be useful.

So, I could take a stab at making an initial ProjectQuadratureFunction, but it probably wouldn't be the most efficient. If you have a preference for what the L2-project should look like let me know. Also, if you're aware of somewhere that's already done this within MFEM that would be a nice starting point for me.